libraries/chain/include/eosio/chain/controller.hpp

```diff
@@ -328,15 +328,16 @@ namespace eosio::chain {
 
          void set_savanna_lib_id(const block_id_type& id);
 
+         // thread-safe
          bool   fork_db_has_root() const;
+         // thread-safe
          size_t fork_db_size() const;
-
-         // thread-safe, applied LIB, fork db root
-         uint32_t last_irreversible_block_num() const;
-         // thread-safe, applied LIB, fork db root
-         block_id_type last_irreversible_block_id() const;
-         // thread-safe, applied LIB, fork db root
-         time_point last_irreversible_block_time() const;
+         // thread-safe, only valid if fork_db_has_root()
+         uint32_t fork_db_root_block_num() const;
+         // thread-safe, only valid if fork_db_has_root()
+         block_id_type fork_db_root_block_id() const;
+         // thread-safe, only valid if fork_db_has_root()
+         time_point fork_db_root_block_time() const;
```

Comment:
I think a better API would be to return a `block_handle` here, even if it involves a `block_ptr` copy.

Reply:
A `signed_block` does not provide an efficient `.id()` call.

Reply:
my bad, I misspoke, the `block_handle` contains the `block_state_ptr` not `signed_block_ptr`.

Reply:
Oh right. Sure might as well match `fork_db_head()`

Reply:
Done.